numpy/_core/defchararray.pyi

```diff
@@ -463,100 +464,132 @@ __all__: list[str]
 @overload
 def equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def equal(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
 @overload
 def not_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def not_equal(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def not_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
 @overload
 def greater_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def greater_equal(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def greater_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
 @overload
 def less_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def less_equal(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def less_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
 @overload
 def greater(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def greater(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def greater(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
 @overload
 def less(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
+def less(x1: B_co, x2: B_co) -> NDArray[np.bool]: ...
+@overload
 def less(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
 
-# String operations
 @overload
-def add(x1: U_co, x2: U_co) -> NDArray[str_]: ...
+def add(x1: U_co, x2: U_co) -> NDArray[np.str_]: ...
+@overload
+def add(x1: B_co, x2: B_co) -> NDArray[np.bytes_]: ...
 @overload
-def add(x1: S_co, x2: S_co) -> NDArray[bytes_]: ...
+def add(x1: S_co, x2: S_co) -> np.ndarray[_Shape, np.dtypes.StringDType]: ...
```

Comment:
The "`TypeVar` with constraints" trick can't be used here unfortunately.

But you could consider using a `TypeAlias` for `np.ndarray[_Shape, np.dtypes.StringDType]` to make this a bit more DRY